go-controller/pkg/util/pod_annotation.go

```diff
@@ -379,3 +389,415 @@ func GetK8sPodAllNetworkSelections(pod *v1.Pod) ([]*nadapi.NetworkSelectionEleme
 	}
 	return networks, nil
 }
+
+// AllocateToPodWithRollbackFunc is a function used to allocate a resource to a
+// pod that depends on the current state of the pod, and possibly updating it.
+// To be used with UpdatePodWithAllocationOrRollback. Implementations can return
+// a nil pod if no update is warranted. Implementations can also return a
+// rollback function that will be invoked if the pod update fails.
+type AllocateToPodWithRollbackFunc func(pod *v1.Pod) (*v1.Pod, func(), error)
+
+// UpdatePodWithRetryOrRollback updates the pod with the result of the
+// allocate function. If the pod update fails, it applies the rollback provided by
+// the allocate function.
+func UpdatePodWithRetryOrRollback(podLister listers.PodLister, kube kube.Interface, pod *v1.Pod, allocate AllocateToPodWithRollbackFunc) error {
+	start := time.Now()
+	var updated bool
+
+	err := retry.RetryOnConflict(OvnConflictBackoff, func() error {
+		pod, err := podLister.Pods(pod.Namespace).Get(pod.Name)
+		if err != nil {
+			return err
+		}
+
+		// Informer cache should not be mutated, so copy the object
+		pod = pod.DeepCopy()
+		pod, rollback, err := allocate(pod)
+		if err != nil {
+			return err
+		}
+
+		if pod == nil {
+			return nil
+		}
+
+		updated = true
+		err = kube.UpdatePod(pod)
+		if err != nil && rollback != nil {
+			rollback()
+		}
+
+		return err
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to update pod %s/%s: %w", pod.Namespace, pod.Name, err)
+	}
+
+	if updated {
+		klog.Infof("[%s/%s] pod update took %v", pod.Namespace, pod.Name, time.Since(start))
+	}
+
+	return nil
+}
+
+// UpdatePodAnnotationWithRetry updates the pod annotation on the pod retrying
+// on conflict
+func UpdatePodAnnotationWithRetry(podLister listers.PodLister, kube kube.Interface, pod *v1.Pod, podAnnotation *PodAnnotation, nadName string) error {
+	updatePodAnnotationNoRollback := func(pod *v1.Pod) (*v1.Pod, func(), error) {
+		var err error
+		pod.Annotations, err = MarshalPodAnnotation(pod.Annotations, podAnnotation, nadName)
+		if err != nil {
+			return nil, nil, err
+		}
+		return pod, nil, nil
+	}
+
+	return UpdatePodWithRetryOrRollback(
+		podLister,
+		kube,
+		pod,
+		updatePodAnnotationNoRollback,
+	)
+}
+
+// ipAllocator is an IP allocation interface that needs to be implemented to use
+// AllocatePodAnnotation
+type ipAllocator interface {
```

Comment:
we already have allocator defined in allocator package, and now we are defining another one in the util package? It seems like we are building more than utility functions here, and I'm afraid its making things more complicated.

Reply:
This type is not exported and you should only care about it if you use the function that needs it. We don't have an equivalent interface exported anywhere else.

But I can certainly define one and export it from the ip allocator package itself.